Implement MongoDB  data connector

krinart opened the pull request:
## 📝 Summary

Support for mongoDB Data Connector with integration tests

Next steps: 
* documentation/examples
* benchmarks
* e2e github workflows

Example:
```
version: v1
kind: Spicepod
name: test-app

datasets:
  - from: mongodb:customer
    name: customer
    params:
        mongodb_connection_string: mongodb://root:root@localhost:27017/tpch?authSource=admin&tls=false
```

## 🔗 Related

https://github.com/datafusion-contrib/datafusion-table-providers/pull/386
https://github.com/spiceai/spiceai/issues/4011

## 🚨 Breaking Changes

N/A

## 📚 Docs

Will update documentation and examples once this PR is merged

krinart committed:
Initial implementation for mongo support

krinart committed:
Fixes + integration tests

krinart committed:
More improvements

krinart committed:
Merge trunk

krinart committed:
Fix tests

krinart committed:
Run cargo fmt

krinart committed:
Fix clippy errors

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Fix cargo fmt

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

mach-kernel committed:
Merge branch 'trunk' into viktor_mongo_support_2

mach-kernel commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,203 @@
+    ParameterSpec::component("sslmode"),
+    ParameterSpec::component("sslrootcert"),
+    ParameterSpec::component("auth_source"),
+    ParameterSpec::component("pool_min")
```
nit: The Rust Driver min pool size is configured to 0 by default, maybe a good idea to match their defaults?

https://www.mongodb.com/docs/drivers/rust/current/fundamentals/performance/#std-label-rust-performance-pool

krinart replied:
Done. Updated both min (0) and max (10) to match

mach-kernel reviewed (dismissed):
Looks good! Gave it a whirl:

```bash
$ mongosh
use testDB
db.users.insertMany([
{ name: "Alice", age: 28, email: "alice@example.com"},
  { name: "Bob", age: 32, email: "bob@example.com" },
  { name: "Charlie", age: 24, email: "charlie@example.com" },
  { name: "Dana", age: 29, email: "dana@example.com" }
])
```

...

```
sql> select * from users;
+-----+---------+--------------------------+---------------------+
| age | name    | _id                      | email               |
+-----+---------+--------------------------+---------------------+
| 28  | Alice   | 6888c78ef8a448c3579e1dc1 | alice@example.com   |
| 32  | Bob     | 6888c78ef8a448c3579e1dc2 | bob@example.com     |
| 24  | Charlie | 6888c78ef8a448c3579e1dc3 | charlie@example.com |
| 29  | Dana    | 6888c78ef8a448c3579e1dc4 | dana@example.com    |
+-----+---------+--------------------------+---------------------+

Time: 0.0188675 seconds. 4 rows.
sql> describe users
  -> ;
+-------------+-----------+-------------+
| column_name | data_type | is_nullable |
+-------------+-----------+-------------+
| age         | Int32     | YES         |
| name        | Utf8      | YES         |
| _id         | Utf8      | YES         |
| email       | Utf8      | YES         |
+-------------+-----------+-------------+
```

krinart committed:
Update default min/max pool size to match mongo driver

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Add direct_connection config param

kczimm commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,204 @@
+                    }
+                    parsed_pool_min.ok()
+                })
+                .unwrap_or(10);
+            let mut pool_max = params
+                .parameters
+                .get("pool_max")
+                .ok()
+                .and_then(|s| {
+                    let pool_max_str = s.expose_secret();
+                    let parsed_pool_max = pool_max_str.parse::<usize>();
+                    if parsed_pool_max.is_err() {
+                        tracing::warn!(
+                            "Invalid pool_max value: {pool_max_str}, using default of 100"
+                        );
+                    }
+                    parsed_pool_max.ok()
+                })
+                .unwrap_or(100);
+
+            if pool_min > pool_max {
+                tracing::warn!(
+                    "pool_min value: {pool_min} is greater than pool_max value: {pool_max}, using default values of 10 and 100"
+                );
+                pool_min = 10;
+                pool_max = 100;
```
Can we put these magic numbers in a single `const` location (i.e. `const DEFAULT_POOL_MIN: usize = 10`), since we refer to them a couple times?

krinart replied:
Makes sense, done.

krinart committed:
Move min/max pool size to constants

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Fix

krinart committed:
Proper constants

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Add time_zone/unnest_depth params

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
fix lint

krinart committed:
Fix lint

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

peasee committed:
Merge branch 'trunk' into viktor_mongo_support_2

peasee commented on .github/workflows/features.yml:
```diff
@@ -197,3 +197,12 @@ jobs:
           CARGO_INCREMENTAL: 0
           CARGO_NET_GIT_FETCH_WITH_CLI: true
+
+      - name: Build with only mongodb`
```
```suggestion
      - name: Build with only mongodb
```

krinart replied:
Done, thanks!

peasee commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,213 @@
+    ParameterSpec::component("sslmode"),
+    ParameterSpec::component("sslrootcert"),
+    ParameterSpec::component("auth_source"),
+    ParameterSpec::component("direct_connection"),
```
Could all of these parameters get descriptions?

krinart replied:
Done, thanks!

peasee commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,213 @@
+                    let parsed_pool_max = pool_max_str.parse::<usize>();
+                    if parsed_pool_max.is_err() {
+                        tracing::warn!(
+                            "Invalid pool_max value: {pool_max_str}, using default of 100"
```
```suggestion
                            "Invalid pool_max value: {pool_max_str}, using default of {DEFAULT_MAX_POOL_SIZE}"
```

krinart replied:
Thanks!

peasee commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,213 @@
+                    let parsed_pool_min = pool_min_str.parse::<usize>();
+                    if parsed_pool_min.is_err() {
+                        tracing::warn!(
+                            "Invalid pool_min value: {pool_min_str}, using default of 10"
```
```suggestion
                            "Invalid pool_min value: {pool_min_str}, using default of {DEFAULT_MIN_POOL_SIZE}"
```

krinart replied:
Thanks!

peasee commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,213 @@
+
+            if pool_min > pool_max {
+                tracing::warn!(
+                    "pool_min value: {pool_min} is greater than pool_max value: {pool_max}, using default values of 10 and 100"
```
```suggestion
                    "pool_min value: {pool_min} is greater than pool_max value: {pool_max}, using default values of {DEFAULT_MIN_POOL_SIZE} and {DEFAULT_MAX_POOL_SIZE}"
```

krinart replied:
Thanks!

peasee commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,213 @@
+                    .insert("pool_max".to_string(), pool_max.to_string().into());
+            }
+
+            let pool = match MongoDBConnectionPool::new(params.parameters.to_secret_map()).await {
```
I think we should check to ensure a user hasn't specified both a `connection_string` and connection details (`host`, `user`, etc).

We should warn that the connection string will be used, and the other parameters ignored.

krinart replied:
done, thanks!

Advayp commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,213 @@
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to create MongoDB connection pool: {source}"))]
```
Are there any steps the user can take to resolve this error? If so, include them in this error message.

krinart replied:
This is actually not used, it just mirrors the same approach as in mysql/postgress. I removed it, thanks!

Advayp commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,213 @@
+        .context(super::UnableToGetReadProviderSnafu {
+            dataconnector: "mongodb",
+            connector_component: ConnectorComponent::from(dataset),
+        })?)
```
```suggestion
        Read::table_provider(
            &self.mongodb_factory,
            dataset.path().into(),
            dataset.schema(),
        )
        .await
        .context(super::UnableToGetReadProviderSnafu {
            dataconnector: "mongodb",
            connector_component: ConnectorComponent::from(dataset),
        })
```
nit: Can we just do this?

krinart replied:
Done, thanks!

krinart committed:
Address comments

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

krinart committed:
Run fmt

krinart committed:
Update cargo.lock

peasee committed:
Merge branch 'trunk' into viktor_mongo_support_2

Advayp commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,249 @@
+
+            let pool = match MongoDBConnectionPool::new(params.parameters.to_secret_map()).await {
+                Ok(pool) => Arc::new(pool),
+                Err(error) => match error {
```
Maybe implement `Into<DataConnectorError>` for `MongoDBError`?

krinart replied:
It seems like [mysql](https://github.com/spiceai/spiceai/blob/trunk/crates/runtime/src/dataconnector/mysql.rs#L173-L209)/[postgres](https://github.com/spiceai/spiceai/blob/trunk/crates/runtime/src/dataconnector/postgres.rs#L103-L140) follow the same pattern (and probably more). 

Maybe it would make sense to create at ticket and implement this logic for all the relevant data connectors to make sure the code is consistent?

peasee commented on crates/runtime/src/dataconnector/mongodb.rs:
```diff
@@ -0,0 +1,249 @@
+                    tracing::warn!(
+                        "Both 'connection_string' and individual connection parameters ({}) were provided. \
+                     The 'connection_string' will be used and the listed parameters will be ignored.",
+                        ignored.join(", ")
```
```suggestion
                        "Both 'connection_string' and individual connection parameters ({parameters}) were provided for the {component}. The 'connection_string' will be used and the listed parameters will be ignored.",
                        parameters = ignored.join(", "),
                        component = params.component
```

peasee reviewed (commented):
Small nit on the warning log wording, but otherwise looks good to me technically!

Could you please update the `# Spec` section of the [MongoDB Data Connector enhancement](https://github.com/spiceai/spiceai/issues/4011) to include details of all of the parameters as well? For example, similar to the [Oracle Data Connector enhancement](https://github.com/spiceai/spiceai/issues/6314).

Once the spec is updated, we can put the issue through UX review and get this PR merged :smile: Thanks!

krinart committed:
Update crates/runtime/src/dataconnector/mongodb.rs

Co-authored-by: peasee <98815791+peasee@users.noreply.github.com>

krinart commented:
@peasee Thanks for the great feedback. I've addressed all the comments and updated enhancement ticket with the list of all parameters.

Please let me know if there's anything else needed from my end to merge this PR. Thanks!

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

peasee reviewed (dismissed):
The enhancement has passed UX review, thanks for your work!

krinart committed:
Update Cargo.toml/lock

krinart committed:
Merge branch 'trunk' into viktor_mongo_support_2

phillipleblanc committed:
Merge branch 'trunk' into viktor_mongo_support_2
